sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/AccessTokenCache.java

```diff
@@ -0,0 +1,165 @@
+                            // still in timeout, do not refresh
+                            tokenRefresh = Mono.empty();
+                        }
+                        // cache hasn't expired, ignore refresh error this time
+                        fallback = Mono.just(cache);
+                    }
+                    return tokenRefresh
+                       .materialize()
+                       .flatMap(signal -> {
+                           AccessToken accessToken = signal.get();
+                           Throwable error = signal.getThrowable();
+                           if (signal.isOnNext() && accessToken != null) { // SUCCESS
+                               logger.info(refreshLog(cache, now, "Acquired a new access token"));
+                               cache = accessToken;
+                               monoProcessor.onNext(accessToken);
+                               monoProcessor.onComplete();
+                               nextTokenRefresh = OffsetDateTime.now().plus(REFRESH_DELAY);
+                               return Mono.just(accessToken);
+                           } else if (signal.isOnError() && error != null) { // ERROR
+                               logger.error(refreshLog(cache, now, "Failed to acquire a new access token"));
+                               nextTokenRefresh = OffsetDateTime.now().plus(REFRESH_DELAY);
+                               return fallback.switchIfEmpty(Mono.error(error));
+                           } else { // NO REFRESH
+                               monoProcessor.onComplete();
+                               return fallback;
+                           }
+                       })
+                       .doOnError(monoProcessor::onError)
+                       .doOnTerminate(() -> wip.set(null));
+                } else if (cache != null && !cache.isExpired() && !forceRefresh) {
+                    // another thread might be refreshing the token proactively, but the current token is still valid
+                    return Mono.just(cache);
+                } else {
+                    // another thread is definitely refreshing the expired token
+
+                    //If this thread, needs to force refresh, then it needs to resubscribe.
+                    if (forceRefresh) {
+                        return Mono.empty();
+                    }
+
+                    MonoProcessor<AccessToken> monoProcessor = wip.get();
+                    if (monoProcessor == null) {
+                        // the refreshing thread has finished
+                        return Mono.just(cache);
+                    } else {
+                        // wait for refreshing thread to finish but defer to updated cache in case just missed onNext()
+                        return monoProcessor.switchIfEmpty(Mono.defer(() -> Mono.just(cache)));
+                    }
+                }
+            } catch (Throwable t) {
+                return Mono.error(t);
```

Comment:
Should we continue attempting to refresh the cache if there is an error?

Reply:
nope, retry policy should kick in.